src/protocol/models.rs

```diff
@@ -50,6 +54,19 @@ impl ProtocolComponent {
     }
 }
 
+pub trait TryFromWithBlock<T> {
+    type Error;
+
+    #[allow(async_fn_in_trait)]
+    async fn try_from_with_block(
+        value: T,
+        block: Header,
+        all_tokens: HashMap<H160, ERC20Token>,
+    ) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
```

Comment:
I have a generic rust question:
Is it normal to name the trait `DoSomething` and then the method `do_something` as well? isn't it redundant? or this is just rust rusting? I'm asking because of `TryFromWithBlock` and `try_from_with_block`